### ISSUE TITLE
Several issues with IValueFormatter

### ISSUE THREAD

**Issue by rcdailey** [fluentassertions/fluentassertions#1527]
### Description

I have a number of issues with the value formatter system right now. To summarize:

- Lack of documentation. I've found nothing that explains how to add a value formatter to Fluent Assertions. Only thing I found was [a very brief section](https://fluentassertions.com/extensibility/#rendering-objects-with-beauty) in the docs with a small example. It doesn't tell me much.
  - Also there is no information on how to properly use `FormatChild`.
- I have a custom object that keep as a value in a dictionary. I never see `CanHandle()` invoked for this type.

### Complete minimal example reproducing the issue

Concerning the custom IValueFormatter, I have mine set up like this right now:

```cs
    public class ProfileDataValueFormatter : IValueFormatter
    {
        public bool CanHandle(object value)
        {
            return value is ProfileData;
        }

        public string Format(object value, FormattingContext context, FormatChild formatChild)
        {
            var profileData = (ProfileData) value;
            return $"{profileData.Ignored}";
        }
    }
```

The `Format()` method right now is completely wrong; my goal is just to see *something* happen, I was planning on flushing it out once I understand more about how custom value formatters work.

`ProfileData` is defined as follows:

```cs
    public class ProfileData
    {
        public List<string> Required { get; init; } = new();
        public List<string> Ignored { get; init; } = new();
        public Dictionary<int, List<string>> Preferred { get; init; } = new();
    }
```

The test case invoking the failed assertion is set up like so (there's a bit of missing context to  understand what is going on, but this is the gist of it):

```cs
            var results = context.ParseWithDefaults(markdown);

            var expectedResults = new Dictionary<string, ProfileData>
            {
                {
                    "Test Release Profile", new ProfileData
                    {
                        Ignored = new List<string> {"added1"},
                        Preferred = new Dictionary<int, List<string>>
                        {
                            {10, new List<string> {"added2", "added3"}}
                        }
                    }
                },
                { "Second Release Profile", new ProfileData()}
            };

            results.Should().Equal(expectedResults);
```

The last line where I do `Should().Equal()` is failing. It prints like so:

```txt
Expected results to be equal to {[Test Release Profile, Trash.Sonarr.ReleaseProfile.ProfileData], [Second Release Profile, Trash.Sonarr.ReleaseProfile...
```

### Expected behavior:

The whole reason I want a custom formatter is because I want something meaningful to print out for `Trash.Sonarr.ReleaseProfile.ProfileData`, which is pretty useless right now for debugging purposes. What would be ideal is to have the members of `ProfileData` printed in a JSON-like format. I want the values in all of the containers of that object to use Fluent's built-in rendering.

I lack the documentation to achieve this goal, but the reason I've filed the bug is not just because of documentation. The main reason is that I believe `CanHandle()` should receive an `object` that can be cast to `ProfileData`, but when I step through the debugger the furthest down it seems to go is `KeyValuePair<string, ProfileData>`. I was expecting one for `string` (which I think I did see) and another call for `ProfileData` as it "walks down" the object tree.

### Actual behavior:

Sort of explained that above... 

### Versions

* Which version of Fluent Assertions are you using? `5.*`
* Which .NET runtime and version are you targeting? E.g. .NET framework 4.6.1 or .NET Core 2.1. `.NET 5`

**Comment by rcdailey** [comment]
Another thing I forgot to mention as far as documentation goes, It never mentioned how to add the formatter into the framework, so I tried this:

```cs
            Formatter.AddFormatter(new ProfileDataValueFormatter());
```

I do this from my `[OneTimeSetup]` method in my NUnit test fixture. I'm not even sure if this is right.

**Comment by dennisdoomen** [comment]
Looking at the docs, I think it's fair what you're asking. Because of your report, I realized this is caused by several issues.

* Normally, FA should be fine formatting the `ProfileData` object as individual properties, but it seems we completely forgot adding a formatter that understands dictionaries. 
* The example doesn't use `FormatChild`
* Apparently we never documented how to add a custom formatter to the `AssertionOptions.Formatters` collection, nor how to use the `[ValueFormatter]` attribute
* Search on the website is broken

Unfortunately, we just introduced a bunch of breaking changes on `develop` (which is going to be part of the upcoming 6.0), so for now, please check out [TaskValueFormatter](https://github.com/fluentassertions/fluentassertions/blob/bde29a65651343bd2386674d5226ab393153e840/Src/FluentAssertions/Formatting/TaskFormatter.cs#L20) or the use of [the attribute here](https://github.com/fluentassertions/fluentassertions/blob/bde29a65651343bd2386674d5226ab393153e840/Tests/FluentAssertions.Specs/Formatting/FormatterSpecs.cs#L795).

**Comment by rcdailey** [comment]
Thank you. What is the specific reason to bring up 6.0? Is it changing how the value formatting is working?

**Comment by dennisdoomen** [comment]
Yes, we had to change the signature of the `IValueFormatter.Handle` method to be able to ensure the rendering never exceeds a certain number of lines for performance reasons. See https://github.com/fluentassertions/fluentassertions/blob/develop/docs/_pages/extensibility.md#rendering-objects-with-beauty
